wallet: return error msg for  "too-long-mempool-chain"

furszy opened the pull request:
Fixes #23144.

We currently return a general "Insufficient funds" from Coin
Selection when we actually skipped unconfirmed UTXOs that
surpassed the mempool ancestors limit.

This PR make the error clearer by returning:
"Unconfirmed UTXOs are available, but spending them creates
a chain of transactions that will be rejected by the mempool"

Also, added an early return from Coin Selection if the sum of
the discarded coins decreases the available balance below the
target amount.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -566,15 +566,18 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, const std::vec
 
     // Add preset inputs to result
-    res->AddInput(preset_inputs);
+    auto resMerged = *res.getObjResult();
```
In c2ec6661cbb70a388dc67d6b4b47f2ab49b99418 "wallet: make SelectCoins return a CallResult."

nit: We prefer snake_case for new variables. (here and elsewhere)

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -557,12 +557,15 @@ CallResult<SelectionResult> SelectCoins(const CWallet& wallet, const std::vector
+            if (auto r8{AttemptSelection(wallet, value_to_select,
+                                         CoinEligibilityFilter(0, 1, std::numeric_limits<uint64_t>::max(), std::numeric_limits<uint64_t>::max(), true /* include_partial_groups */),
+                                         vCoins, coin_selection_params)}) {
+                if (fRejectLongChains) {
```
In c8a43eca7c4081f8ca87be1befbbcdbfdc798a10 "send: return proper error message for too-long-mempool-chain failure."

I'm not a huge fan of doing all of the coin selection computation just to immediately throw away the result. It seems like there should be a better way to do this, e.g. by changing how we fill `vCoins` to be guaranteed to have a solution at this point iff long chain utxos are included.

furszy replied:
hmm yeah, nice point. Will cook something for it.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -651,13 +651,12 @@ static void DiscourageFeeSniping(CMutableTransaction& tx, FastRandomContext& rng
 }
 
-static bool CreateTransactionInternal(
+static OperationResult CreateTransactionInternal(
```
In 30feca7b374256ff62443b2e2f1c65dfa49bb083 "send: move wallet CreateTransaction* flow to use OperationResult/CallResult."

I don't think this needs to be an `OperationResult`. It really can be `CallResult<CTransactionRef>`. The original really should have just returned `CTransactionRef` anyways.

achow101 reviewed (commented):
The last 2 commits should be dropped from this PR and a new PR for each one opened. They are orthogonal to the main changes here.

furszy commented:
Hey, thanks for the review @achow101.

> The last 2 commits should be dropped from this PR and a new PR for each one opened. They are orthogonal to the main changes here.

Sounds good, will move them to their own PRs. I added them here basically because I wasn't totally happy with the extra wasted computation neither, so.. thought on other ways to speed up the process.

---
Will tackle the code styling changes as well 👍.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -83,7 +83,14 @@ TxSize CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *walle
+                    const CAmount& nMaximumAmount,
+                    const CAmount& nMinimumSumAmount,
+                    const uint64_t nMaximumCount,
+                    std::vector<COutput>* vUnconfCoins)
```
In b2846445135f6d6b7902672cb2202dc75088a200 "wallet: add CoinControl filtering capabilities for coins that exceed mempool restrictions."

nit: For new variables, we prefer snake_case, without the prefix for type. So this could be just `unconf_coins`.

achow101 reviewed (commented):
It seems like 0bd751f5a6abfbf97dce94948949c9c95e380a09 "spend: refactor the entire CreateTransaction flow to return a CallResult<CTransactionRef>." could be squashed into 3318ee2cb9e2b372ec9e413fe103e181052b47dc "send: move wallet CreateTransaction* flow to use OperationResult/CallResult."

furszy commented:
kk, done both. You caught me doing some further changes.

Have re-coded the mempool ancestors/descendants cache differently, much cleaner and less intrusive. It's now integrated into the available coin information (the changes are quite small now, it shouldn't require another PR for it but let me know what you think achow101).

Plus cleaned the `CreateTransaction` changes a bit more.

Szatan181 commented:
You can explain more to me what's going on and what exactly you did


wt., 19 kwi 2022, 23:19 użytkownik Matias Furszyfer <
***@***.***> napisał:

> kk, done both. You caught me doing some further changes.
>
> Have re-coded the mempool ancestors/descendants cache differently, much
> cleaner and less intrusive. It's now integrated into the available coin
> information (the changes are quite small now, it shouldn't require another
> PR for it but let me know what you think achow101).
>
> Plus cleaned the CreateTransaction changes a bit more.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/24845#issuecomment-1103184961>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AWTHVFRFNWKAJZX4ES4TCK3VF4PPJANCNFSM5TLGUZWQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

furszy commented:
Aside from achow101 review changes, the first part:
>Have re-coded the mempool ancestors/descendants cache differently, much cleaner and less intrusive. It's now integrated into the available coin information

Was the move from fbc130aa to d9b8ea because now the ancestors and descendants count is available inside `COutput`. Which simplified the changes greatly.

And the second part was an unification for 76aa7d05 of the previous two `emplace_back` calls into a single one inside `AvailableCoins` and moving the ancestors/descendants getter call after the basic checks.

achow101 commented on src/operationresult.h:
```diff
@@ -0,0 +1,44 @@
+
+#include <util/translation.h>
+
+class OperationResult
```
In f198d27ef7f80b6b1c03f388eae216402f1ff424 "introduce OperationResult and CallResult classes"

Since `OperationResult` is only used by `CallResult` now, what do you think about consolidating the two classes?

furszy replied:
I left it because we can make use of it in other functions that only return a bool and the error string without the extra object. Like for example `CConnman::BindListenPort`, `Chain::broadcastTransaction` and `CWallet::SubmitTxMemoryPoolAndRelay`, `BerkeleyEnvironment::Open`, `BerkeleyDatabase::Verify` etc.

I mean, having the class available would let others use it on different PRs if they are touching any similar function signature.

Saying that, if you are still against merging it without a connected use-case, can just remove it and re-introduce it later when we modify any of those functions as well (or.. make an extra small commit that connects it to one of those functions just to present the use-case for it).

achow101 commented on src/operationresult.h:
```diff
@@ -0,0 +1,44 @@
+    explicit operator bool() const { return m_res; }
+};
+
+inline OperationResult ErrorOut(const bilingual_str& error_str)
```
In f198d27ef7f80b6b1c03f388eae216402f1ff424 "introduce OperationResult and CallResult classes"

This function is no longer used.

achow101 commented:
The first 3 commits are fine (besides the comments I left above), but I'm still not sure that this is the correct approach to adding the error message. It seems like it's a ton of added complexity to have the second vector just to be able to say whether the result would have a too long mempool chain.

furszy commented:
> The last 2 commits should be dropped from this PR and a new PR for each one opened. They are orthogonal to the main changes here.

Just dropped the only_spendable `CoinControl` filter option commit. Will open a new PR for it next week.

furszy commented:
Some extra notes to check this week:

- [ ] As fba930d8 + 04d8343e are moving the mempool cs lock from `SelectCoins` to `AvailableCoins` (independently from the extra locking improvements), need to check whether that has any impact on the other `AvailableCoins` functions callers (not only the `CreateTransaction*` flow that already know that is fine). Because now they can be locking the mempool cs when before they were not doing it. So.. in other words, sanity check over the mutexes locking order for callers that aren't part of the flow being changed here.

------

Then another possible path that I'm thinking if there is no quorum over the changes here is move the new mempool info cache and coin control filtering capabilities to a follow-up PR and add here a straightforward, but slower and redundant, approach like:

 Loop over the available coins inside `CreateTransactionInternal` if the `SelectCoins` call fails to basically get the mempool info on-demand and check if there is any long-chain coin before returning the generic "Insufficient funds" error.

Which will correct the base problem that this PR is aiming to fix without adding the other stuff. Then, moving forward, can push the other improvements on independent PRs so they can be discussed atomically.

achow101 commented:
Instead of having all of this mempool policy lookup stuff, what if we instead just had `GroupOutputs` produce all of the sets of `OutputGroup`s that we will end up using during coin selection? Then each filtered set of groups would be passed in for each `AttemptSelection`. This way, we could also pre-calculate the total value available for a given set of groups and skip `AttemptSelection`s when there is clearly not enough value for that group set. This would also let us get the too-long-mempool-chain error as we could see whether the group set that allows the long mempool chains would have enough value to potentially reach a solution.

A downside of this approach would be the memory use. This would end up consuming a lot more memory as we would be making all of the group sets for each `SelectCoins` rather than for each `AttemptSelection`. However we may be able to work around that by using references, but we have to be careful as some of the coin selection algorithms modify the group set that they are given.

furszy commented:
Small update here:
Now that #25218 was merged, the next step for this is #25685 which restructures the coin selection area further.

Still, meanwhile, I will be pushing the new outputs grouping changes in the coming days. Need to add proper test coverage for it first, and.. will try to decouple them from #25685, moving forward :tractor:

furszy committed:
wallet: return error msg for too-long-mempool-chain failure

We currently return "Insufficient funds" which doesn't really
describe what went wrong; the tx creation failed because of
a long-mempool-chain, not because of a lack of funds.

Also, return early from Coin Selection if the sum of the
discarded coins decreases the available balance below the
target amount.

furszy committed:
test: add wallet too-long-mempool-chain error coverage

S3RK commented on src/wallet/spend.cpp:
```diff
@@ -711,8 +728,18 @@ util::Result<SelectionResult> AutomaticCoinSelection(const CWallet& wallet, Coin
+        // Let's check if we have long-chain of unconfirmed UTXOs in the not accepted groups
+        for (const auto& group : not_accepted_groups) {
+            if (group.m_ancestors >= max_ancestors)
+                return util::Result<SelectionResult>({_("Unconfirmed UTXOs are available, but spending them creates a chain of transactions that will be rejected by the mempool")});
```
Just because these UTXOs are available it doesn't mean that including them would cover required amount.

For example let's imaging you have two coins in a wallet:
- 48 BTC with 6 confirmation
- 1 BTC with long unconfirmed ancestry

If you try to send 50 BTC you should get "insufficient funds" error, but IIUC after this PR you'll get "Unconfirmed UTXOs are available..." which is unrelated.

furszy replied:
yeah good, fixed.
also, pushed an early return if the sum of the not accepted groups decreases the available amount below the target (we simply don't have enough balance after applying the filters, so no need to run the selection algos).

furszy commented:
Thanks @S3RK!, feedback tackled.

We now check, and return early, if there is no enough available balance after filtering groups and before running the selection algos.

murchandamus reviewed (commented):
ACK f3221d373a8623fe4808e00c7a92fbb6e0d6419b

achow101 commented:
ACK f3221d373a8623fe4808e00c7a92fbb6e0d6419b

S3RK commented on src/wallet/spend.cpp:
```diff
@@ -692,7 +708,24 @@ util::Result<SelectionResult> AutomaticCoinSelection(const CWallet& wallet, Coin
+        CAmount total_unconf_long_chain = 0;
+        for (const auto& group : discarded_groups) {
+            total_discarded += group.GetSelectionAmount();
+            if (group.m_ancestors >= max_ancestors || group.m_descendants >= max_descendants) total_unconf_long_chain += group.GetSelectionAmount();
```
1) I think this should respect `fRejectLongChains` flag. If user explicitly asked us to ignore long-chain constraints, then we should respect that.

2) If `fRejectLongChains` flag is set (the default) then the most lenient filter has `max_ancestors/2` and `max_descendants/2`. So this condition could never be hit.

In summary I don't yet understand how one can hit a message described in #23144

furszy replied:
> 1. I think this should respect `fRejectLongChains` flag. If user explicitly asked us to ignore long-chain constraints, then we should respect that.

It's being respected:
If the user asked to ignore long-chain constraint (`fRejectLongChains=false`) the last filter include the coins, which means that they aren't discarded, so the `discarded_groups` vector is empty, so `total_unconf_long_chain` is zero, so the error is never hit.

> 2. If `fRejectLongChains` flag is set (the default) then the most lenient filter has `max_ancestors/2` and `max_descendants/2`. So this condition could never be hit.

At the contraire, the unconfirmed coins are inside the `available_coins` struct (the user sees them as spendable), and because of the not inclusion of the last filter (due the `fRejectLongChains=true` flag), they are excluded from the created output groups and included inside the discarded groups vector, which is the one used to hit this condition.

> In summary I don't yet understand how one can hit the error described in #23144

Thinks in this way:

If the wallet reject long chains (default), the process discard coins that the user sees as spendable and tries to find a result with the remaining ones. Returning an "Insufficient Funds" if it's not found.
The "discarded groups" vector basically represents that. Coin Selection excluded coins seen as spendable by the user.

S3RK replied:
Yes you're right, I somehow managed to confuse myself about the filters. Thanks for patient explanation.

Still something doesn't sit right. So if a coin is discarded because of `max_ancestors` we give this error message, but if it's discarded by the filter `max_ancestors/2` we just say insufficient funds when the coins are there they just don't meet the filters. There are also other reasons when the coins are discarded, for example unsafe utxos. Even you enable unsafe utxos, we don't allow any chains at all. Why give detailed error message in one case and not the other? Maybe we should just count all discarded coins and generalize the error message a bit?

furszy replied:
> Still something doesn't sit right. So if a coin is discarded because of max_ancestors we give this error message, but if it's discarded by the filter max_ancestors/2 we just say insufficient funds when the coins are there they just don't meet the filters

no, those coins will be added to the last non-customizable filter which has `max_ancestors-1` and `max_descendants-1`.

> There are also other reasons when the coins are discarded, for example unsafe utxos. Even you enable unsafe utxos, we don't allow any chains at all

The unsafe UTXO filter is just redundant, we are currently filtering unsafe UTXO inside the coins fetching process (`AvailableCoins`).
In other words, when `m_include_unsafe_inputs=false` no unsafe UTXO arrives at Coin Selection.

> Why give detailed error message in one case and not the other? Maybe we should just count all discarded coins and generalize the error message a bit?

The process is already counting all discarded coins and returning early if their exclusion make the available amount go below the target. It's currently returning an "Insufficient Funds", but could say something different too.

I made the too-long-mempool-chain error an special case because this error, alike the untrusted coins that the user can see in a different balance (the unconfirmed/untrusted one), is hidden to the user as it depends on inner mempool policies.
It's not enough for the user to set a flag to spend the unsafe coins, it needs to set another one to craft the tx with an unconfirmed output that surpasses the max descendants policy.

S3RK replied:
I realised my mistake. You can resolve this and ignore my comments. I'll go over the code one more time before I ack

S3RK commented on src/wallet/spend.cpp:
```diff
@@ -692,7 +708,24 @@ util::Result<SelectionResult> AutomaticCoinSelection(const CWallet& wallet, Coin
+
+        // Check if we still have enough balance after applying filters (some coins might be discarded)
+        CAmount total_discarded = 0;
+        CAmount total_unconf_long_chain = 0;
```
q: do I understand correctly with current filters `total_discarded` is always equal to `total_unconf_long_chain`?

furszy replied:
Nop. Could also have zero conf change (or send-to-self) if the wallet's `m_spend_zero_conf_change` flag is disabled. In other words, unconfirmed coins accepted by the mempool limits (ancestors/descendants < 25).

S3RK commented:
Code review ACK f3221d373a8623fe4808e00c7a92fbb6e0d6419b
